### PR TITLE
add test for parse magnet link

### PR DESCRIPTION
**Pull request by sarp**
This PR adds tests for parse magnet link stage

I will update `course_definition.yml` and `tester_definition.go` in a separate PR at the end

### CODE REVIEW

File: internal/stage_magnet_parse.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	}
+
+	magnetLinks := []MagnetLink {
+		{
+		   UrlEncoded: "magnet:?xt=urn:btih:c77829d2a77d6516F88cd7a3de1a26abcbfab0db&dn=codercat.gif&tr=http%3A%2F%2Fbittorrent-test-tracker.codecrafters.io%2Fannounce",
+		   Trackers: []string { "http://bittorrent-test-tracker.codecrafters.io/announce" },
+		   InfoHash: "c77829d2a77d6516F88cd7a3de1a26abcbfab0db",
+	   },
+	   {
+		UrlEncoded: "magnet:?xt=urn:btih:c77829d2a77d6516F88cd7a3de1a26abcbfab0db",
```
Comment (rohitpaulk) [inline internal/stage_magnet_parse.go:23]: Do we necessarily have to test this case? Doesn't look like we're going to test any such "tracker-less" links in later stages. Can we just make all tests here have at least the info hash and tracker URL?

File: internal/stage_magnet_parse.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+func testParseMagnetLink(stageHarness *test_case_harness.TestCaseHarness) error {
+	type MagnetLink struct {
+		UrlEncoded string
+		Trackers []string
+		InfoHash string
+	}
+
+	magnetLinks := []MagnetLink {
+		{
+		   UrlEncoded: "magnet:?xt=urn:btih:c77829d2a77d6516F88cd7a3de1a26abcbfab0db&dn=codercat.gif&tr=http%3A%2F%2Fbittorrent-test-tracker.codecrafters.io%2Fannounce",
```
Comment (rohitpaulk) [inline internal/stage_magnet_parse.go:18]: We need some level of randomization for both the info hash and tracker URL here. One could pass this stage by just hardcoding both - and often people do this just to ensure they have the "format" correct, and then intend to implement the actual parsing after .

Comment (rohitpaulk) [reply internal/stage_magnet_parse.go:18]: ("Randomization" here could just be including three different cases with different values - don't necessarily mean truly random values)

**Review by rohitpaulk — APPROVED** [review]

File: internal/stage_magnet_parse.go (hunk 1 above)
Comment (sarp) [reply internal/stage_magnet_parse.go:23]: Yep, if our goal is to ensure that people can pass all stages, we don't need to test this case
My goal was to cover edge cases people might see in the real world, like no trackers, multiple trackers, malformed link, extra parameters etc. because otherwise we get the criticism that it feels like a toy project and not the real thing ...

File: internal/stage_magnet_parse.go (hunk 2 above)
Comment (sarp) [reply internal/stage_magnet_parse.go:18]: Good point, will update

File: internal/stage_magnet_parse.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		Trackers: []string { ""},
+		InfoHash: "c77829d2a77d6516F88cd7a3de1a26abcbfab0db",
+	   },
+	}
+
+	logger := stageHarness.Logger
+	executable := stageHarness.Executable
+
+	for _, link := range(magnetLinks) {
+		logger.Infof("Running ./your_bittorrent.sh magnet_parse %s", escape(link.UrlEncoded))
```
[the remark this answers is not in the document]
Comment (sarp) [reply internal/stage_magnet_parse.go:33]: Yep, will do
